src/sp/transport/udp/udp.c

```diff
@@ -964,7 +965,17 @@ udp_pipe_recv(void *arg, nni_aio *aio)
 		nni_mtx_unlock(&ep->mtx);
 		return;
 	}
-
+	//If there are messages in the queue, retrieve the messages from the queue during reception 
+	//without waiting for the udp_recv_data operation to complete.
+	if(!nni_lmq_empty(&p->rx_mq)){
+		nni_lmq_get(&p->rx_mq, &msg);
+		nni_aio_set_msg(aio, msg);
+		nni_aio_completions_add(
```

Comment:
Thank you for working on this.

You shouldn't do this particular step however.

The purpose of the completion queue is to allow us to batch aio completions and run those completions without reacquiring the lock, and should only be done on inbound receive path whereas this is being called downwards.

The right thing to do here is simply finish the aio inline.   You can do that by calling:

``nni_aio_finish(aio, 0, nni_aio_count(aio));` 

You can (and should) do that *after* dropping the lock.  

Also,  it would be good if you would use clang-format to fix the whitespace.  For example `//` is always followed by a space, and we insert spaces elsewhere.  (I will make the style suggestions inline.)